Tweak Input focusBorderColor API

mattrothenberg opened the pull request:
As per https://github.com/chakra-ui/chakra-ui/issues/91, this PR proposes a few adjustments to the `Input` component's API.

One discrepancy that I've noted throughout the codebase is the ambiguity of the `color` prop. In some cases, users are expected to pass a "partial" color string, e.g., `teal`. In other cases, users are expected to pass a fully-formed color string that corresponds with a key in the `colors` object that powers the UI, e.g., `teal.500`.

I'm partial to the latter approach, but ultimately I care more about consistency. To that end, I've tweaked the `Input` API accordingly.

1. Change the default value of `focusBorderColor` from `blue` to `blue.500`, so as to avoid unnecessary string interpolation in the input style generator.
2. Applied the value of the `focusBorderColor` prop to the `flushed` variant, which was otherwise ignored before.

I do have a few open questions, though –

1. Should we expose a prop for `errorColor`? It feels odd to allow users to customize the focus state, but not the error state.
2. I noticed that we were treating `focusBorderColor` differently when the dark mode was applied (dropping the color from `500` to `300`). Is there any reason for this? Should we defer this opinion to users, allowing them to pass whatever `focusBorderColor` they want as a function of whether the app is in dark mode or not?

Hope this is helpful, thanks for letting me share my thoughts!

segunadebayo commented:
Hi @mattrothenberg ,

I really like your feedback and I agree when you say that the `color` prop is quite inconsistent and I'm looking to find ways to make it consistent. Just allowing the user to pass any color they want is better.

I appreciate the changes to the Input component.

To answer your questions:
1. It's a good idea to provide the `errorBorderColor` prop to make it possible to style the border when the input is invalid. We can default it to `red.500`

2. I changed the `focusBorderColor` in dark mode following the Material Design Spec for dark theme. Dark mode is typically a low-light, dark background scenario, and to provide enough color contrast, it's advised to use a lighter color value, hence the switch from `500` to `300`.

To make it more declarative, I think we'll have to allow the user to determine whichever color they want to use in light and dark mode. Thanks for the suggestion.

I'll look into this more thoroughly and ensure more uniform usage of any `color` prop. For now, kindly add the `errorBorderColor` prop to the input, update the type definition and I'll merge the PR.

mattrothenberg commented:
@segunadebayo I've addressed your feedback in https://github.com/chakra-ui/chakra-ui/pull/92/commits/92525f4f8b2f24db49cdb4c3d0b9575b544235a9.

Thanks again for your work on this awesome library!

PierreAndreis commented:
Bump on this. It doesn't make sense to have a powerful theme when we can't change the focus outline color

with-heart commented:
I'm not sure why this popped up in my notifications. But while I'm here:

@allcontributors please add @mattrothenberg for code and examples
